lib/wibox/widget/base.lua

```diff
@@ -321,6 +330,372 @@ function base.place_widget_at(widget, x, y, width, height)
+--
+-- Code:
+--    
+--    mywibox[s] : setup {
+--        {
+--            id     = "second",
+--            widget = wibox.widget.textbox
+--        },
+--        {
+--            id     = "third",
+--            widget = wibox.widget.textbox
+--        },
+--        id     = "first",
+--        layout = wibox.layout.fixed.horizontal,
+--    }
+--    
+--    mywibox[s].first.second:set_markup("changed!")
+--    mywibox[s]:get_children_by_id("third")[1]:set_markup("Also changed!")
+--    
+--
+--
+-- **Example 7 :** Extending the system
+--
+-- This system is very flexible. Each section attribute (the entries with string
+-- keys) is directly linked to the layout or widget API. When setting the
+-- imaginary `myproperty`, it will first check if `set_myproperty` exist. If it
+-- doesn't, it will check if there is a `myproperty` method. Finally, it will
+-- just set the `mywidget.myproperty` directly in case it is used later or
+-- catched by a lua `metatable` (operator overload).
+--
+-- Code:
+--    
+--    -- "Monkeypatch" a new function to wibox.widget.textbox to add vicious
+--    -- extension support
+--    function wibox.widget.textbox.vicious(w, args)
+--        local f = unpack or table.unpack -- Lua 5.1 compat
+--        vicious.register(w, f(args))
+--    end
+--
+--    mywibox[s] : setup {
+--        {
+--            vicious = {vicious.widgets.cpu, "CPU: $1%", 3},
+--            widget  = wibox.widget.textbox
+--        },
+--        layout = wibox.layout.fixed.horizontal,
+--    }
+--    
+--
+-- In this example, the system is extended to that the popular
+-- [Vicious](http://awesome.naquadah.org/wiki/Vicious) extension module can be
+-- used directly in the layout declaration. This example will update the textbox
+-- every 3 seconds to set the CPU usage.
+--
+--
+-- **Example 8 :** Handling sections
+--
+-- The system is allow sections to be defined externally, then composed into
+-- the final layout declaration. Here is an example re-using example #4:
+--
+-- Code:
+--    
+--    local circle = {
+--        fit    = function(self, context, width, height)
+--            return height, height -- A square taking the full height
+--        end,
+--        draw   = function(self, context, cr, width, height)
+--            cr:set_source_rgb(1, 0, 0) -- Red
+--            cr:arc(height/2, height/2, height/2, 0, math.pi*2)
+--            cr:fill()
+--        end,
+--        layout = wibox.widget.base.make_widget,
+--    }
+--
+--    -- 3 circle
+--    mywibox[s] : setup {
+--        circle,
+--        circle,
+--        circle,
+--        layout = wibox.layout.align.horizontal
+--    }
+--    
+--    -- This can be done instead
+--    local three_circle = {layout = wibox.layout.align.horizontal}
+--    for i=1, 3 do
+--        table.insert(three_circle, circle)
+--    end
+--    
+--    mywibox[s] : setup (three_circle)
+--    
+--
+-- @param args An array containing the widgets disposition
+function base.widget:setup(args)
```

Comment:
I hope someone will read it eventually, that took a while to write ;)

But it is also the base of the unit test I will add. This (hopefully) cover all case, and all the examples are copy-pastable into the default `rc.lua` to replace `mywibox[s]:set_widget()`, so a user can try them right away (onlt the first one is quite useless, it the widgets are `nil`, I removed them as it made the example too big)